### PR TITLE
fix(ui): settle drag selections after pointer release

### LINKED ISSUE #2400: test(e2e): eliminate the quote selection drag timing race
<details open>
<summary><strong>English</strong></summary>

## Symptom

The `e2e_shard (1/2)` check can fail in `e2e/quote-companion.spec.ts` while asserting that `.maka-quote-actions` stays hidden during a drag selection:

```text
Error: expect(locator).toBeHidden() failed
Locator: locator('.maka-quote-actions')
Expected: hidden
Received: visible
```

Observed in [CI run 31154080906, job 92789739992](https://github.com/maka-agent/maka-agent/actions/runs/31154080906/job/92789739992). The other 55 tests in the shard passed, as did the other E2E shard, Windows baseline, typecheck, Runtime Host, workspace, and headless checks.

## Cause

The test drives five mouse movements with a 90 ms driver-side pause after each movement and expects every movement to extend the 350 ms selection-settle timer. Once the pointer moves beyond the available text, however, the browser may stop emitting `selectionchange` because the selection itself no longer changes. Runner scheduling can then let the existing settle timer expire before `mouse.up()`, making the quote layer legitimately visible under the current event model while the test still expects it to be hidden.

This makes the assertion depend on driver and runner timing instead of an observable synchronization boundary. The failing PR did not modify `use-message-selection-quote.ts`, `chat-view.tsx`, or this E2E.

## Expected fix

- define whether an unchanged selection with the pointer still held down is considered settled
- make the implementation and E2E use that explicit boundary
- avoid fixed driver-side delays as the correctness oracle
- preserve coverage that the layer does not flash during an actively changing selection

</details>

<details>
<summary><strong>简体中文</strong></summary>

## 现象

`e2e_shard (1/2)` 可能在 `e2e/quote-companion.spec.ts` 中失败：测试在拖选文本期间断言 `.maka-quote-actions` 保持隐藏，但实际浮层已经可见。

本次出现在 [CI run 31154080906, job 92789739992](https://github.com/maka-agent/maka-agent/actions/runs/31154080906/job/92789739992)。该 shard 的其余 55 项测试、另一个 E2E shard、Windows baseline、typecheck、Runtime Host、workspace 和 headless 检查均通过。

## 原因

测试通过 5 次鼠标移动和每次 90 ms 的 driver 侧等待模拟拖选，并假设每次移动都会重新开始 350 ms 的 selection-settle timer。但指针越过可选文本后，selection 本身可能不再变化，浏览器也就不再发送 `selectionchange`。当 CI runner 调度稍慢时，已有 timer 会在 `mouse.up()` 前到期；按照当前事件模型，quote layer 会正常出现，而测试仍要求它隐藏。

因此该断言依赖 driver 与 runner 时序，而不是明确、可观察的同步边界。触发失败的 PR 没有修改 `use-message-selection-quote.ts`、`chat-view.tsx` 或这条 E2E。

## 预期修复

- 明确“鼠标仍按下但 selection 不再变化”是否算 settled
- 让实现与 E2E 使用同一个显式边界
- 不再用固定的 driver 侧延时作为正确性判断依据
- 保留“selection 仍在变化时浮层不能闪现”的覆盖

</details>

### PR DESCRIPTION
**Pull request by me2seeks**
## Summary

The quote affordance currently treats 350 ms without a `selectionchange` as a settled drag. Once the pointer moves past the available text, the browser may stop changing the Selection while the button is still held, so the layer can appear before the gesture ends.

This makes primary pointer release the commit boundary for drag selection. Selection changes while that pointer is held clear the old affordance but do not arm the timer; pointerup then starts the existing settle path. Keyboard and programmatic selections keep the same delay, while unchanged clicks, another pointer, pointer cancellation, and window blur cannot resurrect a dismissed quote.

Pointer tracking starts only inside a turn that can own a quote. Controls elsewhere in the scroll surface, including the quote actions, therefore cannot clear the quote they are trying to consume.

The E2E drives a deterministic pointer/selection event boundary through the real transcript DOM without driver-side timing sleeps. A deterministic UI boundary test owns the timing policy.

Fixes #2400

<details>
<summary>中文对照</summary>

引用浮层目前把“350ms 内没有 `selectionchange`”当成拖选已经结束。当指针越过可选文本后，浏览器可能不再改变 Selection，但鼠标仍处于按下状态，因此浮层会在手势结束前出现。

本 PR 将主指针释放设为拖选的提交边界。指针按住期间的选区变化只会清除旧浮层，不会启动 timer；pointerup 后才进入原有 settle 路径。键盘和程序化选区仍保留相同延迟，同时未改变选区的点击、其他 pointer、pointercancel 与窗口失焦都不能让已关闭的引用浮层重新出现。

pointer tracking 只会从可拥有引用的 turn 内部开始，因此滚动区域中的其他控件（包括引用操作按钮本身）不会在消费引用前把它清掉。

E2E 改为在真实 transcript DOM 中确定性地驱动 pointer/selection event 边界，不再依赖 driver 侧 sleep；确定性的 UI boundary 单测负责时序规则。

</details>

## Verification

- `npm --workspace @maka/ui test` — 265 tests passed
- `npm --workspace @maka/desktop run build:with-deps`
- `npx biome check` on all changed source, test, and E2E files
- Focused Electron E2E passed under a local Xvfb display, including `CI=1` visible-window mode

## Checklist

- [x] Tests cover the change and fail without it
- [x] Lint, format, typecheck and the affected non-Electron suite pass locally
- [x] Focused Electron E2E passed on Xvfb; GitHub merge-sha run pending

Does this PR entail a change in behavior?

- [x] Yes — a drag-selected quote can appear only after its pointer is released
- [ ] No

### CODE REVIEW

**Review by Astro-Han — COMMENTED** [review]
Thanks for making pointer release the explicit selection boundary — the normal in-window path is clear and the small state machine is easy to follow.

**[P2]** One window-boundary case still looks open: a mouse drag can leave the Electron window and be released without the document receiving `pointerup` and without the window necessarily blurring. Because the gesture does not capture the pointer, `activePointerId` can then remain set; later keyboard or programmatic selection changes keep hiding the quote without scheduling settlement until another matching pointer release happens.

Could the owning turn capture the pointer and also settle or cancel on `lostpointercapture`? A focused real-DOM/Electron test for releasing outside the window would help cover the listener wiring—the current head only contains the pure boundary tests, since the earlier quote E2E is no longer present.

Thanks again — the original in-window timing race does look correctly addressed.

**Comment by me2seeks** [comment]
Addressed the window-boundary case: the owning turn now captures the pointer and settles on `lostpointercapture`, with Electron coverage for releasing beyond the viewport. Would appreciate another look when you have a chance.

File: packages/ui/src/use-message-selection-quote.ts [hunk 1]
```diff
@@ -100,16 +168,95 @@ export function useMessageSelectionQuote(
...
+    function onPointerCancel(event: PointerEvent): void {
+      if (event.pointerId === capturedPointerId) clearCaptureOwner();
+      gesture.cancelPointerSelection(event.pointerId);
+    }
+
+    function onLostPointerCapture(event: Event): void {
+      // Normal pointerup already settles first, making this a no-op. When the
+      // owner loses capture without pointerup, this is the only trustworthy
+      // end boundary left for the final selection.
+      const pointerId = (event as PointerEvent).pointerId;
```
Comment (Astro-Han) [inline packages/ui/src/use-message-selection-quote.ts:257]: **[P2] Could capture loss remain a cancellation boundary here?**

My reading is that `lostpointercapture` does not necessarily mean the pointer was released: explicit release, capture transfer, owner removal, or pointer lock can fire it while the pointer remains active. Calling `endPointerSelection()` then schedules settlement after an earlier `selectionchange`, so the quote could appear while the button is still held—the invariant this PR is establishing.

A clean option would be to route terminal events through one helper with an explicit disposition:

```ts
finishGesture(pointerId, 'commit'); // pointerup
finishGesture(pointerId, 'cancel'); // pointercancel or capture loss
```

The helper could detach the owner listener, clear the active gesture, and schedule only for `commit && changed`. The E2E’s `lostpointercapture` marker could then go away: on the normal path, `onPointerUp` removes this production listener before implicit capture loss fires, so that marker does not exercise this branch. The outside-viewport `pointerup` and final quote assertions are the valuable coverage.

If Electron guarantees that every loss reaching this listener represents a physical release, or if another invariant makes early settlement impossible, please push back—I may be missing that platform-specific guarantee.

File: packages/ui/src/use-message-selection-quote.ts [hunk 2]
```diff
@@ -100,16 +168,95 @@ export function useMessageSelectionQuote(
...
+        event.button !== 0 ||
+        !targetNode ||
+        !turnOwner ||
+        !root.contains(turnOwner) ||
+        (interactiveOwner !== null && turnOwner.contains(interactiveOwner)) ||
+        findEnclosingTurnId(
+          targetNode as unknown as QuoteScopeNode,
+          root as unknown as QuoteScopeNode,
+        ) === null
+      ) {
```
Comment (Astro-Han) [inline packages/ui/src/use-message-selection-quote.ts:223]: **[P2] Could the gesture and capture share one accepted owner?**

`isPrimary` is scoped per pointer type, so mouse, touch, and pen can each have a primary pointer concurrently. `beginPointerSelection()` ignores a second pointer while one is active, but this caller still replaces `captureOwner` and `capturedPointerId` and captures that pointer. The logical gesture can therefore remain owned by pointer A while the DOM metadata moves to pointer B, leaving A’s listener outside the normal cleanup path.

Returning an accepted/rejected result from `beginPointerSelection()` and establishing capture only on acceptance would be the smallest fix. If you want to remove the duplicated authority entirely, one nullable record is another option:

```ts
type ActiveSelectionGesture = {
  pointerId: number;
  owner: Element;
  changed: boolean;
};
```

Then `pointerdown` either creates the complete record or changes nothing, `selectionchange` marks it changed, and the terminal path consumes it.

If the supported Electron input model rules out concurrent primary pointer types here, please push back with that constraint; in that case the smaller guard may be sufficient.

**Review by Astro-Han — APPROVED** [review]
Approving this revision. The original timing race is fixed at the right boundary: `selectionchange` remains the authority for what changed, the owning pointer release commits a drag selection, and pointer capture preserves the release path beyond the viewport. The focused Electron coverage and the full CI run are green.

I left two non-blocking P2 follow-ups around the capture lifecycle. They appear to share one underlying issue: the logical gesture identity and the DOM capture identity can currently diverge. One clean direction would be a single nullable `{ pointerId, owner, changed }` record and a shared terminal helper with explicit `commit` and `cancel` outcomes. That is a suggestion rather than a prescribed implementation—if Chromium/Electron provides a stronger ordering guarantee here, or there is a smaller way to preserve the same invariant, please push back with that evidence.

I would keep both levels of coverage, while simplifying them around the behavior they actually prove: table-test the meaningful state transitions, retain the outside-viewport Electron `pointerup` path, and avoid treating the browser’s normal post-`pointerup` capture-loss event as coverage of the production fallback.

Review disclosure: Codex assisted with an adversarial analysis of the changed pointer state machine, browser event semantics, and tests, including an independent AI review pass. I (Astro-Han) reviewed the current head, verified the relevant event ordering and CI state, and made the approval and severity decisions.

File: packages/ui/src/use-message-selection-quote.ts (hunk 1 above)
Comment (me2seeks) [reply packages/ui/src/use-message-selection-quote.ts:257]: Fixed in a9bbf460e. `lostpointercapture` now cancels the accepted gesture through the shared terminal helper, so it clears state without scheduling quote settlement. The focused boundary test covers the changed-selection cancellation path.

File: packages/ui/src/use-message-selection-quote.ts (hunk 2 above)
Comment (me2seeks) [reply packages/ui/src/use-message-selection-quote.ts:223]: Fixed in a9bbf460e. The gesture boundary now owns one `{ pointerId, owner, changed }` record and `beginPointerSelection()` returns whether it accepted the pointer. Capture and its listener are installed only after acceptance, so a second primary pointer cannot replace the active owner.

**Review by Astro-Han — APPROVED** [review]
Thanks for consolidating the gesture and capture lifecycle in `a9bbf46`. I reviewed the final incremental commit, and it addresses both follow-ups from the previous approval.

The accepted gesture now owns one `{ pointerId, owner, changed }` record, a second primary pointer cannot replace that owner, and the shared terminal path distinguishes commit on `pointerup` from cancellation on `pointercancel` or `lostpointercapture`. Removing the normal capture-loss assertion from the Electron test also keeps that test focused on the real outside-viewport release behavior.

The final head is cleanly mergeable, and all CI checks are green. I did not find any remaining reproducible P0–P3 issues. Approved and ready to merge.

### AI-assisted review disclosure

Codex performed the incremental source and test inspection. I, Astro-Han, reviewed the evidence and made the final review decision.
